### PR TITLE
Add frontend, backend, and timing to access log

### PR DESCRIPTION
**Pull request by tooda02**
This PR enhances the access log to include frontend and backend names and total time.  Log format is:

    127.0.0.1 - - [14/Apr/2016:16:30:45 +0000] "GET /test HTTP/1.0" 200 31 "" "ApacheBench/2.3" 334110 "frontend" "http://127.0.0.1:8081" 56.846528ms

The PR adds the last four tokens:

* reqid (a count of the total number of requests since Traefik startup)
* frontend name (derived from the backend name via a map built from the current configuration)
* backend URL (assigned by oxy)
* elapsed time

This PR replaces PR #308, primarily to change the owner of the origin repo from tooda02 to CiscoCloud.  This repo is also based on the latest Traefik master (as of 4/19/2016)

#### Overview of Update
1. middlewares/logger.go is substantially revised to gather and report the new information.  It creates the reqid and stores it in a new custom header, X-Traefik-Reqid that can be accessed when subsequent middleware methods call function saveBackendNameForLogger(). The reqid provides a context for storing names. Note that the HTTP request pointer cannot be used as a context key (making Gorilla context unsuitable) because oxy replaces the HTTP request with a shallow copy when assigning a backend. The logger deletes the custom header before returning to its caller.

2. middlewares/saveBackend.go runs after oxy has assigned a backend and calls saveBackendNameForLogger() to report it to the logger, which saves it in a concurrent map.

#### Testing
Three tests are provided with the PR

* A unit test for logger.go in middlewares/logger_test.go.   Run with `go test` from the middlewares directory.

* A functional test in examples/accessLog.  This test starts three http servers and provides a curl command to send requests to them so that the access.log update can be verified. To run it:
  1. Run script examples/accessLog/runExample.sh from the Traefik root directory.
  2. Run command `curl http://127.0.0.1:80/test{1,2,2}` to send some requests.
  3. Run command `tail log/access.log` to verify the access.log was written correctly.

* A benchmark stress test derived from the functional test above.  This uses [Apache ab](https://httpd.apache.org/docs/2.4/programs/ab.html), which must be pre-installed.   Use script examples/accessLog/runAb.sh to run it.  This script supports several options for setting the number of connections, number of requests, and total time of the test.  Use command `runAb.sh --help` to list them.

#### Response to Comments
@emilevauge [commented](https://github.com/containous/traefik/pull/215#issuecomment-199936524) on the previous PR to make some requests:

* could you squash your commits?
  _The repo has been recreated and has only two commits, with all source updates in the first._
* rather than a bash script test, I would prefer integration or unit tests of your code ;)
  _The PR now includes a unit test in addition to the script_
* I would rename routes to saveFrontend to stay homogeneous
  _routes.go is no longer part of the PR, as recent changes to server.go broke its functionality.  The PR now builds a backend-to-frontend map and derives the frontend name from the backend URL.  FYI, the server change that broke saveFrontend was to add an unnamed subrouter to the router with the path.  mux.RouteMatch() matches both, but returns only the unnamed subrouter._
* Did you make benchmarks on traefik using that (using ab, or wrk), to test high loads?
_The PR includes a benchmark script using ab.  I used it to run a one-hour test with 100 concurrent connections, which showed no memory leaks or concurrency issues._

### CODE REVIEW

File: server.go [hunk 1]
```diff
@@ -35,7 +35,10 @@ import (
 	"github.com/streamrail/concurrent-map"
 )
 
-var oxyLogger = &OxyLogger{}
+var (
+	oxyLogger           = &OxyLogger{}
+	backend2FrontendMap map[string]string
```
Comment (emilevauge) [inline server.go]: Not sure this map is thread-safe.

**Comment by emilevauge** [comment]
Hi @tooda02 thanks a lot :+1: 
If it's possible, it would be great to have some integration test for that instead of manual tests ( those in `examples)`. Do you think it's possible?
I made one comment.
I'll try to test it with `-race` later.
/cc @vdemeester

**Comment by tooda02** [comment]
@emilevauge Thanks for the review.  I've made a commit addressing your comments:

* I removed the global map (it wasn't being used anyway due to a minor bug further down)
* I removed the map from middlewares/logger and just save the pointer passed from server.go.  This should eliminate any possible concurrency issues
* I've added an integration test

File: middlewares/logger.go [hunk 2]
```diff
@@ -25,17 +63,134 @@ func NewLogger(file string) *Logger {
...
+	backend2FrontendMap = newMap
+}
+
 func (l *Logger) ServeHTTP(rw http.ResponseWriter, r *http.Request, next http.HandlerFunc) {
 	if l.file == nil {
 		next(rw, r)
 	} else {
-		handlers.CombinedLoggingHandler(l.file, next).ServeHTTP(rw, r)
+		reqid := strconv.FormatUint(atomic.AddUint64(&reqidCounter, 1), 10)
+		log.Debugf("Starting request %s: %s %s %s %s %s", reqid, r.Method, r.URL.RequestURI(), r.Proto, r.Referer(), r.UserAgent())
```
Comment (emilevauge) [inline middlewares/logger.go]: Could you remove this log?

File: middlewares/logger.go [hunk 3]
```diff
@@ -25,17 +63,134 @@ func NewLogger(file string) *Logger {
...
+		r.Header[loggerReqidHeader] = []string{reqid}
+		defer deleteReqid(r, reqid)
+		frontendBackendLoggingHandler{reqid, l.file, next}.ServeHTTP(rw, r)
 	}
 }
 
-// Close closes the logger (i.e. the file).
+// Delete a reqid from the map and the request's headers
+func deleteReqid(r *http.Request, reqid string) {
+	log.Debugf("Ending request %s", reqid)
```
Comment (emilevauge) [inline middlewares/logger.go]: Could you remove this log?

**Comment by emilevauge** [comment]
Hi @tooda02, thanks for your corrections. I made 2 minor comments to remove logs.
I tested with `-race` and saw no issue during a benchmark.
LGTM (once rebased) :)
/cc @vdemeester

**Comment by tooda02** [comment]
Rebased with master and removed debug logs as requested.

File: examples/accessLog/exampleHandler.go [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+	if *help || len(name) == 0 || port <= 0 {
+		usage()
+	}
+	http.HandleFunc("/", handler)
+	fmt.Printf("%s: Listening on :%d...\n", name, port)
+	if er := http.ListenAndServe(fmt.Sprintf(":%d", port), nil); er != nil {
+		fmt.Printf("%s: Error from ListenAndServe: %s", name, er.Error())
+		os.Exit(1)
+	}
+	fmt.Printf("%s: How'd we get past listen and serve???\n", name)
```
Comment (vdemeester) [inline examples/accessLog/exampleHandler.go]: :smile_cat:

File: integration/access_log_test.go [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+	c.Assert(err, checker.IsNil)
+	defer cmd.Process.Kill()
+	defer os.Remove("access.log")
+	defer os.Remove("traefik.log")
+
+	time.Sleep(500 * time.Millisecond)
+
+	// Verify Traefik started OK
+	if traefikLog, err := ioutil.ReadFile("traefik.log"); err != nil {
+		c.Assert(err.Error(), checker.Equals, "")
```
Comment (vdemeester) [inline integration/access_log_test.go]: I don't really get this part but, I would think you could use `IsNil` checker instead, like : 

```go
traefikLog, err := ioutil.ReadFile("traefik.log")
c.Assert(err, checker.IsNil)
c.Assert(len(traefikLog), checker.Equals, 0)
```

File: integration/access_log_test.go [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+
+	checker "github.com/vdemeester/shakers"
+)
+
+// AccessLogSuite
+type AccessLogSuite struct{ BaseSuite }
+
+func (s *AccessLogSuite) TestAccessLog(c *check.C) {
+	// Ensure working directory is clean
+	os.Remove("access.log")
```
Comment (vdemeester) [inline integration/access_log_test.go]: I would prefer we use a `ioutil.TempDir` folder that is always empty when running the test, but that would mean have a way to "parametrized" the configuration file (we do this elsewhere already). I could make a follow-up PR to do that tough :wink:.

File: integration/access_log_test.go [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+	// Make some requests
+	_, err = http.Get("http://127.0.0.1:8000/test1")
+	c.Assert(err, checker.IsNil)
+	_, err = http.Get("http://127.0.0.1:8000/test2")
+	c.Assert(err, checker.IsNil)
+	_, err = http.Get("http://127.0.0.1:8000/test2")
+	c.Assert(err, checker.IsNil)
+
+	// Verify access.log output as expected
+	if accessLog, err := ioutil.ReadFile("access.log"); err != nil {
```
Comment (vdemeester) [inline integration/access_log_test.go]: Same as above, I don't really get the assertion just below.

```go
accessLog, err := ioutil.ReadFile("access.log")
c.Assert(err, checker.IsNil)
// […]
```

File: integration/access_log_test.go [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+		c.Assert(err.Error(), checker.Equals, "")
+	} else {
+		lines := strings.Split(string(accessLog), "\n")
+		count := 0
+		for i, line := range lines {
+			if len(line) > 0 {
+				count++
+				if tokens, err := shellwords.Parse(line); err != nil {
```
Comment (vdemeester) [inline integration/access_log_test.go]: And same here :angel:

File: integration/access_log_test.go (hunk 6 above)
Comment (vdemeester) [inline integration/access_log_test.go]: Or we could use `ioutil.TempDir`, run traefik from it and then go back to where we were.

File: middlewares/logger.go [hunk 9]
```diff
@@ -1,18 +1,56 @@
 package middlewares
 
+/*
+Middleware Logger writes each request and its response to the access log.
```
Comment (vdemeester) [inline middlewares/logger.go]: This comment should be on the Logger struct (to appear correctly in `godoc`).

File: middlewares/saveBackend.go [hunk 10]
```diff
@@ -0,0 +1,24 @@
+package middlewares
+
+/*
+Middleware saveBackend sends the backend name to the logger.
```
Comment (vdemeester) [inline middlewares/saveBackend.go]: Same here, this comment should be on the Logger struct (to appear correctly in `godoc`).

**Comment by vdemeester** [comment]
@tooda02 Thanks, great work overall :+1:
Few comments though, that I could take care of in a follow-up PR if you don't feel like it :wink: 

>  examples/k8s.namespace.sh 100755 → 100644
Is this intended ? :angel:

**Comment by tooda02** [comment]
@vdemeester Thanks for the review.  I've made a commit addressing your comments, except that I left the traefik.log assertion:

```go
if len(traefikLog) > 0 {
    fmt.Printf("%s\n", string(traefikLog))
    c.Assert(len(traefikLog), checker.Equals, 0)
}
```
as it was.   The benefit of this over the simpler

```go
c.Assert(len(traefikLog), checker.Equals, 0)
```

is that if the assertion fails, the unexpected log error is reported to build output (instead of just the length of it), making it easier to diagnose the issue.   The gymnastics around err were for a similar reason - to make sure that the error text was printed when the assertion failed.  This was needed with another assert package I've used, but not here.

**Comment by vdemeester** [comment]
@tooda02 Thanks !
For the assertion, you could actually do `c.Assert(len(traefikLog), checker.Equals, 0, check.Commentf(string("traefik.log")))` but anyway, that's fine by me as is :wink: 

LGTM 🐸

**Pull request merged**